### DISCUSSION TITLE
Convert all comment h1 tags into h2 tags automatically (maybe even h3?)

### DISCUSSION THREAD

**Discussion by benhalpern** [forem/forem#17053] [Feature Requests]
For creating _posts_, we typically deal with accessibility concerns by alerting the author to changes they _should probably make_, and while I wouldn't be opposed to also automatically changing H1 tags to H2 in posts and not bother with the nudge, I think we could have much more latitude to do this _in comments_, where I think there aren't the same expectations for precision and ownership of style the way there is for posts.

I.e. comments belong more directly to the community than posts, and we should have more freedom to make choices on the user's behalf.

And with that preamble, I think disallowing H1 tags in comments seems like a win on a number of dynamics:

- Of course, accessibility
- SEO: Pages which are fast and loose with H1 tags do not perform well. Semantic consistency, even in user-generated content should be a goal.
- UX consistency: A non-screen-reader end user probably doesn't care _that much_ about the semantics, but tightening up what can be rendered in comments will be a small help in overall consistency.

***

This is a pretty fast change without many downstream concerns. I don't think we'd want to necessarily bother _re-saving all historical comments_, via data update script, though we could manually look in to backfilling them on DEV after. I'd advocate for not backfilling as part of the PR itself as to not over-complicate what could be a very quick win.

**Comment by aitchiss** [comment]
I like this in theory, and if it helps SEO then I'm all for it. But we should be aware that there is scope for us to cause a _different_ accessibility issue unless we cascade the change by incrementing every other heading level in the same comment, in that the document outline created by the headings would fall out of step. Maybe that's not such a huge concern for comments though, given they're not really used for long-form, sectioned, content 🤔 

Following below is just an illustration of what I mean by this:

```
# My section about dev tools

## Network tools

## HTML inspector tools
```

The above conveys a relationship between the top heading and the lower two, marking them as subsections (some screen readers even surface headings like this as nested lists to convey that structure):

- My section about dev tools
  - Network tools
  - HTML inspector tools

If we don't cascade the change to all subsequent headings, we'd be changing that to:

```
## My section about dev tools

## Network tools

## HTML inspector tools
```
i.e.

- My section about dev tools
- Network tools
- HTML inspector tools

We'd fix the a11y issue of duplicate H1s but we've introduced another accessibility problem, as we now have no relationship conveyed between the comment's top-level section and its sub-sections. I'm not sure which is the worse accessibility failing to be honest 😄  I guess given comments don't tend, by nature, to include lots of nested sections it's probably on balance still better UX to change H1 -> H2, even if sectioning falls out of step as a result.

**Comment by aitchiss** [comment]
> (maybe even h3?)

H3 would actually be more semantically correct I think, as our comments section begins with an H2. But then we would need to also convert all H2s -> H3 I think.

**Comment by jeremyf** [comment]
If we were to put each comment within an Article/Section, I believe that would mean they can have their own header structure.

See https://html.spec.whatwg.org/multipage/sections.html#the-h1,-h2,-h3,-h4,-h5,-and-h6-elements about semantic equivalence of the H-tags

**Comment by aitchiss** [comment]
> If we were to put each comment within an Article/Section, I believe that would mean they can have their own header structure.

Unfortunately I don't believe that has been implemented by any browsers 😞  See [some context here](https://css-tricks.com/how-to-section-your-html/#aa-no-browser-supports-the-document-outline-algorithm), and from the MDN docs:

> The HTML specification includes the concept of an outline formed by the use of `<section>` elements. If this were implemented it would enable the use of multiple `<h1>` elements, giving user agents—including screen readers—a way to understand that an `<h1>` nested inside a defined section is a subheading. This functionality has never been implemented; therefore it is important to use your headings to describe the outline of your document.

(it would still be beneficial for accessibility to have comments wrapped in `article` elements, just for slightly different reasons)

**Reply by jeremyf** [reply]
Thanks for the clarification and I wish semantics were enough…but alas implementation details always muddy semantics.
